test/Microsoft.AspNetCore.Authentication.WsFederation.Test/WsFederationTest.cs

```diff
@@ -0,0 +1,293 @@
+
+            // Send an invalid token and verify that the token is not honored
+            var kvps = new List<KeyValuePair<string, string>>();
+            kvps.Add(new KeyValuePair<string, string>("wa", "wsignin1.0"));
+            kvps.Add(new KeyValuePair<string, string>("wresult", File.ReadAllText(@"ValidToken.xml")));
+            kvps.Add(new KeyValuePair<string, string>("suppressWctx", "true"));
+            var response = await httpClient.PostAsync(server.BaseAddress + "signin-wsfed", new FormUrlEncodedContent(kvps));
+
+            Assert.Equal(HttpStatusCode.Found, response.StatusCode);
+
+            var request = new HttpRequestMessage(HttpMethod.Get, response.Headers.Location);
+            var cookies = SetCookieHeaderValue.ParseList(response.Headers.GetValues(HeaderNames.SetCookie).ToList());
+            foreach (var cookie in cookies)
+            {
+                if (cookie.Value.HasValue)
+                {
+                    request.Headers.Add(HeaderNames.Cookie, new CookieHeaderValue(cookie.Name, cookie.Value).ToString());
+                }
+            }
+            response = await httpClient.SendAsync(request);
+
+            // Did the request end in the actual resource requested for
+            Assert.Equal(WsFederationDefaults.AuthenticationScheme, await response.Content.ReadAsStringAsync());
+        }
+
+        [Fact]
+        public async Task InvalidTokenIsRejected()
+        {
+            var builder = new WebHostBuilder()
+                .ConfigureServices(ConfigureAppServices)
+                .Configure(ConfigureApp);
+            var server = new TestServer(builder);
+
+            var httpClient = server.CreateClient();
+
+            // Verify if the request is redirected to STS with right parameters
+            var response = await httpClient.GetAsync("/");
+            var queryItems = QueryHelpers.ParseQuery(response.Headers.Location.Query);
+
+            // Send an invalid token and verify that the token is not honored
+            var kvps = new List<KeyValuePair<string, string>>();
+            kvps.Add(new KeyValuePair<string, string>("wa", "wsignin1.0"));
+            kvps.Add(new KeyValuePair<string, string>("wresult", File.ReadAllText(@"InvalidToken.xml")));
+            kvps.Add(new KeyValuePair<string, string>("wctx", queryItems["wctx"]));
+            response = await httpClient.PostAsync(queryItems["wreply"], new FormUrlEncodedContent(kvps));
+
+            // Did the request end in the actual resource requested for
+            Assert.Equal("AuthenticationFailed", await response.Content.ReadAsStringAsync());
+        }
+
+        private void ConfigureAppServices(IServiceCollection services)
+        {
+            services.AddAuthentication(sharedOptions =>
+            {
+                sharedOptions.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+                sharedOptions.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
```

Comment:
Don't need this if its the same as the DefaultScheme?

Reply:
We do today due to the 2.0.x patch bug.